server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -266,10 +266,29 @@ func newStatsRecorder(env environment.Env, openChannels *sync.WaitGroup, onStats
 func (r *statsRecorder) Enqueue(ctx context.Context, beValues *accumulator.BEValues) {
 	persist := &PersistArtifacts{}
 	if !*disablePersistArtifacts {
-		testOutputURIs := beValues.TestOutputURIs()
-		persist.URIs = make([]*url.URL, 0, len(testOutputURIs))
-		persist.URIs = append(persist.URIs, beValues.BuildToolLogURIs()...)
-		persist.URIs = append(persist.URIs, testOutputURIs...)
+		buildToolLogs := beValues.BuildToolLogURIs()
+		failedTestOutputs := beValues.FailedTestOutputURIs()
+		passedTestOutputs := beValues.PassedTestOutputURIs()
+
+		capacity := len(buildToolLogs) + len(failedTestOutputs) + len(passedTestOutputs)
+		if capacity > accumulator.MaxPersistableTestArtifacts {
+			capacity = accumulator.MaxPersistableTestArtifacts
+		}
+		persist.URIs = make([]*url.URL, 0, capacity)
+
+		persist.URIs = append(persist.URIs, buildToolLogs...)
+		persist.URIs = append(persist.URIs, failedTestOutputs...)
+		if len(persist.URIs) < accumulator.MaxPersistableTestArtifacts {
+			remain := accumulator.MaxPersistableTestArtifacts - len(persist.URIs)
+			if remain < 0 {
+				remain = 0
+			}
+			if len(passedTestOutputs) <= remain {
+				persist.URIs = append(persist.URIs, passedTestOutputs...)
+			} else {
+				persist.URIs = append(persist.URIs, passedTestOutputs[:remain]...)
+			}
+		}
```

Comment:
Since we're now limiting the passed and failed artifacts to at most 1K each, it seems fine to just persist both lists as-is instead of limiting their combined length (if you agree and wind up making this change then don't forget to revert `MaxPersistableTestArtifacts` back to being private)

Reply:
Im ok with that idea. Sent a patch to do just that.

Im a bit afraid of the increased load on object storage at the end. But let's see if we run into an actual limit.